lib/internal/process.js

```diff
@@ -197,38 +197,68 @@ function setupSignalHandlers() {
   // Load events module in order to access prototype elements on process like
   // process.addListener.
   const signalWraps = {};
+  const oneshots = {};
 
   function isSignal(event) {
     return typeof event === 'string' && lazyConstants()[event] !== undefined;
   }
 
-  // Detect presence of a listener for the special signal types
-  process.on('newListener', function(type, listener) {
-    if (isSignal(type) &&
-        !signalWraps.hasOwnProperty(type)) {
-      const Signal = process.binding('signal_wrap').Signal;
-      const wrap = new Signal();
+  function startSignalWrap(type, isOneShot) {
+    if (signalWraps[type]) {
+      signalWraps[type].close();
+      delete signalWraps[type];
+    }
 
-      wrap.unref();
+    const Signal = process.binding('signal_wrap').Signal;
+    const wrap = new Signal();
+    wrap.unref();
+
+    wrap.onsignal = function() { process.emit(type); };
+
+    const signum = lazyConstants()[type];
+    const err = wrap.start(signum, isOneShot);
+    if (err) {
+      wrap.close();
+      const errnoException = require('util')._errnoException;
+      throw errnoException(err, 'uv_signal_start');
+    }
 
-      wrap.onsignal = function() { process.emit(type); };
+    signalWraps[type] = wrap;
+  }
 
-      const signum = lazyConstants()[type];
-      const err = wrap.start(signum);
-      if (err) {
-        wrap.close();
-        const errnoException = require('util')._errnoException;
-        throw errnoException(err, 'uv_signal_start');
+  // Detect presence of a listener for the special signal types
+  process.on('newListener', function(type, listener) {
+    if (isSignal(type)) {
+      if (listener.oneshot) {
+        if (oneshots[type])
+          ++oneshots[type];
+        else
+          oneshots[type] = 1;
       }
 
-      signalWraps[type] = wrap;
+      if (!signalWraps.hasOwnProperty(type)) {
```

Comment:
I don't think the `hasOwnProperty()` is needed since it's not possible for signal names to clash with properties inherited from `Object.prototype`.